---
title: "HT transcription units"
author: "Claire Rioualen"
date: '`r Sys.Date()`'
output:
 html_document:
   fig_caption: yes
   highlight: zenburn
   self_contained: yes
   theme: cerulean
   toc: yes
   toc_depth: 5
   toc_float: yes
   number_sections: no
   code_folding: hide


---
<style type="text/css">

body{ /* Normal  */
      font-size: 14px;
  }
</style>


```{r setup, include = FALSE}
setwd("/Users/rioualen/Google Drive/Work/Ecoli_mapping_project/Feature_set")
ht_collection_dir <- "~/Dropbox/HT-Collections"

knitr::opts_knit$set(root.dir = ht_collection_dir)
knitr::opts_chunk$set(echo = TRUE, include = TRUE, warning = FALSE, message = FALSE, cache = FALSE, eval = TRUE, fig.width = 12, fig.height = 9, fig.align = 'center')
```


```{r libs-n-functions}
library(ggplot2)
library(dplyr)
library(DT)
library(gridExtra)
library(readxl)
library(tidyr)

## Pending proper installation
devtools::load_all("~/Desktop/Git/EcoliGenes")

concat_uniq <- function(x, sep = ","){
  paste0(unique(na.omit(x)), collapse = sep)
}

check_dup <- function(df){
  test_dup <- df %>%
    mutate(coords = paste0(start, "_", stop, "_", strand)) %>%
    group_by(coords) %>%
    summarise(count = n(), 
              across(where(is.character), concat_uniq, sep = "_"),
              across(where(is.numeric), concat_uniq, sep = "_"),
              .groups = 'drop') %>%
    arrange(desc(count)) %>%
    filter(count>1) 
  test_dup
}

## wrapper function to remove commented rows from excel sheets
read_excel_cm <- function(file, sheet, comment.char = "#"){
  tab <- read_excel(file, sheet, col_names = F)
  skip_rows <- grep(paste0("^", comment.char), unlist(c(tab[,1])))
  if(length(skip_rows) > 0) tab <- read_excel(file, sheet, skip = max(skip_rows))
  return(tab)
}

```


# Transcription units

## Format 

Here, a **dataset** is defined by a combination of **PMID and experimental condition**. 

TUs are collected from a variety of sources and processed to produce uniform datasets in a bed-like format, with the following fields:

* **chromosome:** NC_000913.3
* **start:** Left genomic position
* **stop:**  Right genomic position
* **id:** Unique TU ID 
* **length:** TU length
* **strand:** TU strand reported by authors
* **term_type:** depending on methodology
* **gene_number:** number of genes entirely contained in TU
* **genes:** bnumbers of genes entirely contained in TU
* **pseudo:** 1 if TU contains pseudo genes, else 0
* **phantom:** 1 if TU contains phantom genes, else 0

## Metadata

```{r metadata-file}
metadata_file <- "DatasetCollection-TUs-HT_v0.3.xlsx"
```

* Metadata file version used: **`r metadata_file`**

```{r read-metadata}
## Read metadata table, filter out datasets that don't have a condition associated
ht_tu_metadata <- read_excel_cm(paste0(ht_collection_dir, "/coleccion_TUs/metadata/", metadata_file), sheet = "TUs_v0.2") %>%
  dplyr::rename(dataset_id = "Dataset ID",
                pmid = "PMID",
                author = "Authors",
                condition = "Growth Conditions Experimental",
                file = "Dataset File Name") %>%
  dplyr::select(dataset_id, pmid, author, condition) %>%
  dplyr::filter(!is.na(condition))

## Make dataset list from table
ht_tu_dataset_list <- setNames(split(ht_tu_metadata, seq(nrow(ht_tu_metadata))), ht_tu_metadata$dataset_id)
```

```{r display-metadata}
## Display datasets
DT1 <- DT::datatable(ht_tu_metadata, options = list(dom = '', pageLength = 20, autoWidth = TRUE, columnDefs = list(list(width = '200px', targets = c(3, 4)))))
DT1
```


## Datasets per author {.tabset}

This part is customized depending on the originally published files. Since those files are formatted each in their own way, this part semi manual and should be modified carefully.

It generates one table per dataset, with the following columns: start, stop, strand, term_type.

### Conway

Reference article: [Conway et al., 2014](https://doi.org/10.1128/mbio.01442-14)

Conditions:

* Exponential phase

Notes:

* Genome coordinates were converted from version U00096.2 to version U00096.3 [here](https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12) 

```{r conway-tu-ht}
## Code executed just once, in order to export promoters coordinates and convert them through Ecocyc
##---------------------------------------------------------------------------------
# file <- ht_tu_dataset_list$TU0003$file
# read_excel(paste0(ht_collection_dir, "/coleccion_TUs/author_files/", file), sheet = "table S4 TUs", skip = 1) %>%
#   dplyr::select(TU, STRAND) %>%
#   tidyr::separate(TU, c("start", "stop"), sep = ":") %>%
#   dplyr::mutate(start = as.numeric(gsub("^[A-Z]{1,2}-{1}", "", start)),
#                 stop = as.numeric(gsub("^[A-Z]{1,2}-{1}", "", stop))) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TUs/author_files/intermediate_files/Conway_TU_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)

## Conversion Conway_TU_parsed.tsv -> Conway_TU_parsed_converted.tsv

TU0003 <- read.delim(paste0(ht_collection_dir, "/coleccion_TUs/author_files/intermediate_files/Conway_TU_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::distinct() %>%
  dplyr::rename(strand = STRAND) %>%
  dplyr::mutate(term_type = "TTS")

```

### Ju 

Reference article: [Ju et al., 2019](https://doi.org/10.1038/s41564-019-0500-z)

Conditions:

* exponential phase
* stationary phase

```{r ju-tu-ht}
##---------------------------------------------------------------------------------
# file <- ht_tu_dataset_list$TU0001$file
file <- "Ju et al_Supplementary Table 3.xlsx"
TU0001_TU0002 <- read_excel(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/", file), skip = 1)  %>%
  dplyr::rename(start = Boundary_of_transcript_unit...1, 
                stop = Boundary_of_transcript_unit...2, 
                strand = Direction_of_transcript_unit) %>%
  dplyr::mutate(term_type = "TTS") %>%
  dplyr::select(-Number_of_covered_annotated_genes, -Name_of_covered_Genes, -Length, -...7)

##---------------------------------------------------------------------------------
TU0001 <- TU0001_TU0002 %>%
  dplyr::filter(!is.na(detected_in_log_phase)) %>%
  dplyr::select(-contains("detected"))

TU0002 <- TU0001_TU0002 %>%
  dplyr::filter(!is.na(detected_in_stationary_phase)) %>%
  dplyr::select(-contains("detected"))
```

### Yan

Reference article: [Yan et al., 2018](10.1038/s41467-018-05997-6)

Conditions:

* Rich (rich growth medium)
* M9 (minimal growth medium)

Notes:

* Two types of TU termination are reported for each condition: detected TTS or longest read end. They are then grouped in 2 files according to growth conditions

```{r yan-tu-ht}
##---------------------------------------------------------------------------------
TU0004 <- read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/M9_RegulonDB_TU_definedEnd.txt"), stringsAsFactors = F) %>% 
  dplyr::mutate(term_type = "TTS") %>% 
  rbind.data.frame(read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/M9_RegulonDB_TU_longestRead.txt"), stringsAsFactors = F) %>% 
                     dplyr::mutate(term_type = "long")) %>%
  dplyr::rename(start = V1, stop = V2, strand = V3) %>%
  dplyr::select(start, stop, strand, term_type)

##---------------------------------------------------------------------------------
TU0005 <- read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/Rich_RegulonDB_TU_definedEnd.txt"), stringsAsFactors = F) %>% 
  dplyr::mutate(term_type = "TTS") %>% 
  rbind.data.frame(read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/Rich_RegulonDB_TU_longestRead.txt"), stringsAsFactors = F) %>% 
                     dplyr::mutate(term_type = "long")) %>%
  dplyr::rename(start = V1, stop = V2, strand = V3) %>%
  dplyr::select(start, stop, strand, term_type)

```


## Uniformization

This part assumes that the specificities of each individual dataset were dealt with previously. What it does:

* Generate unique IDs for TUs
* Generate additional columns (TU length, TU genes, TU gene number, flag columns for pseudo genes and phantom genes)
* Write one file per dataset

Notes:

* Only the genes that are entirely contained in the TUs are taken into account, regardless of what was reported by the authors in the original datasets.
* The step that consists in getting bnumbers from genes that are contained in each TU is a little slow.

```{r write-tu-files}
tu_dir <- paste0(ht_collection_dir, "/coleccion_TUs/uniformized_data")
datasets_list <- list()

## Process all datasets automatically
for(ds_id in names(ht_tu_dataset_list)) {
  df <- get(ds_id) %>%
    dplyr::arrange(start) %>%
    dplyr::mutate(id = paste0("TU_", ds_id, "_", dplyr::row_number()),
                  chromosome = "NC_000913.3") %>%
    dplyr::rowwise() %>%
    dplyr::mutate(length = stop - start + 1,
                  genes = EcoliGenes::what_genes(start, stop, strand)) %>%
    tidyr::separate_rows(genes, sep = ",") %>%
    dplyr::mutate(pseudo = EcoliGenes::is_pseudogene(genes),
                  phantom = EcoliGenes::is_phantomgene(genes)) %>%
    dplyr::group_by(id) %>%
  	dplyr::summarise(across(where(is.character), concat_uniq),
  	                 across(where(is.numeric), min),
  	                 gene_number = ifelse(genes == "", 0, n())) %>%
    dplyr::arrange(start) 


  write.table(df %>% dplyr::select(chromosome, start, stop, id, length, strand, term_type, gene_number, genes, pseudo, phantom),
              file = paste0(tu_dir, "/", ds_id, ".tsv"),
              sep = "\t", quote = F, col.names = T, row.names = F)
  
  assign(ds_id, df)
  
  datasets_list[[ds_id]] <- df

}

```

* Write a summary file for TU datasets

```{r write-tu-table}
ht_tu_stats <- ht_tu_metadata %>%
  dplyr::rowwise() %>%
  dplyr::mutate(tu_number = nrow(get(dataset_id))) %>%
  dplyr::select(dataset_id, pmid, author, condition, tu_number)

write.table(ht_tu_stats, file = paste0(tu_dir, "/TU_datasets.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)
```

## Results

### Summary of datasets

```{r display-tu-table}
DT2 <- DT::datatable(ht_tu_stats, options = list(dom = '', pageLength = 20, autoWidth = TRUE, columnDefs = list(list(width = '200px', targets = c(3, 4))))) %>% DT::formatRound('tu_number', digits=0)
DT2

total_tu <- nrow(bind_rows(datasets_list))

```

Total objects: `r total_tu`

### Distributions

#### Number of TUs per dataset

```{r fig-stats-tu-1}
ht_tu_stats_legends <- ht_tu_stats  %>%
  dplyr::mutate(first_author = str_split(author, ",")[[1]][1])

ggplot(ht_tu_stats_legends, aes(x = dataset_id, y = tu_number, fill = first_author)) +
  geom_col(size = 2) +
  # scale_fill_manual(values = RColorBrewer::brewer.pal(length(unique(ht_tu_stats$first_author)), "Pastel2")) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text = element_text(size = 14)) +
  ylab("Number of TUs") +
  xlab("Dataset ID") +
  labs(title = "Number of TUs per dataset")
  
```

#### Number of genes per TU

```{r fig-stats-tu-2, fig.height = 24}
tu_gene_number_plots <- list()
for(ds_id in names(ht_tu_dataset_list)) {
  author  <- ht_tu_dataset_list[[ds_id]]$author
  condition  <- ht_tu_dataset_list[[ds_id]]$condition
  
  df <- get(ds_id)
  
  gg <- ggplot(df, aes(x = gene_number)) + 
    geom_bar() +
    theme(text = element_text(size = 14)) +
    xlim(c(-0.5, 20)) +
    ylab("Number of TUs") +
    xlab("Number of entire genes per TU") +
    labs(title = ds_id, subtitle = paste0("Author: ", author, ", condition: ", condition))
  tu_gene_number_plots[[ds_id]] <- gg
}
do.call("grid.arrange", c(tu_gene_number_plots, nrow = length(ht_tu_dataset_list)))
```

