---
title: "HT promoters"
author: "Claire Rioualen"
date: '`r Sys.Date()`'
output:
 html_document:
   fig_caption: yes
   highlight: zenburn
   self_contained: yes
   theme: cerulean
   toc: yes
   toc_depth: 5
   toc_float: yes
   number_sections: no
   code_folding: hide


---
<style type="text/css">

body{ /* Normal  */
      font-size: 14px;
  }
</style>


```{r setup, include = FALSE}
setwd("/Users/rioualen/Google Drive/Work/Ecoli_mapping_project/Feature_set")
ht_collection_dir <- "~/Dropbox/HT-Collections"

knitr::opts_knit$set(root.dir = ht_collection_dir)
knitr::opts_chunk$set(echo = TRUE, include = TRUE, warning = FALSE, message = FALSE, cache = FALSE, eval = TRUE, fig.width = 12, fig.height = 9, fig.align = 'center')
```

```{r libs-n-functions}
library(ggplot2)
library(dplyr)
library(DT)
library(gridExtra)
library(readxl)
library(tidyr)

## Pending proper installation
devtools::load_all("~/Desktop/Git/EcoliGenes")

concat_uniq <- function(x, sep = ","){
  paste0(unique(na.omit(x)), collapse = sep)
}

check_dup <- function(df){
  test_dup <- df %>%
    mutate(coords = paste0(start, "_", stop, "_", strand)) %>%
    group_by(coords) %>%
    summarise(count = n(), 
              across(where(is.character), concat_uniq, sep = "_"),
              across(where(is.numeric), concat_uniq, sep = "_"),
              .groups = 'drop') %>%
    arrange(desc(count)) %>%
    filter(count>1) 
  test_dup
}
```

# Transcription start sites

## Format 

Here, a **dataset** is defined by a combination of **PMID and experimental condition**. 

TSSs are collected from a variety of sources and processed to produce uniform bed-like dataset files with the following fields: 

* **chromosome:** NC_000913.3
* **start:** Left genomic position. If not reported, will be the same as `pos_1`
* **stop:**  Right genomic position. If not reported, will be the same as `pos_1`
* **id:** Unique TSS ID 
* **pos_1:** TSS +1 position reported by authors
* **strand:** TSS strand reported by authors

## Metadata

```{r metadata-file}
# metadata_file <- "DatasetCollection-TSS-HT_v0.3.xlsx"
metadata_file <- "DatasetCollection-TSS-HT_v0.5.xlsx"
```

* Metadata file version used: **`r metadata_file`**

```{r read-metadata}
## Read metadata table, filter out datasets that don't have a condition associated
ht_tss_metadata <- read_excel(paste0(ht_collection_dir, "/coleccion_TSS/metadata/", metadata_file), sheet = "DATASET", skip = 1) %>%
  dplyr::rename(dataset_id = "Dataset ID",
                pmid = "PMID",
                author = "Authors",
                condition = "Growth Conditions Experimental",
                file = "Dataset File Name") %>%
  dplyr::select(dataset_id, pmid, author, condition) #%>%
  # dplyr::filter(!is.na(condition))

## Make list from table
ht_tss_dataset_list <- setNames(split(ht_tss_metadata, seq(nrow(ht_tss_metadata))), ht_tss_metadata$dataset_id)
```

```{r display-metadata}
## Display datasets
DT1 <- DT::datatable(ht_tss_metadata, options = list(dom = '', pageLength = 20, autoWidth = TRUE, columnDefs = list(list(width = '200px', targets = c(3, 4)))))
DT1
```

## Datasets per author {.tabset}

### Conway

Reference article: [Conway et al., 2014](https://doi.org/10.1128/mbio.01442-14)

Conditions:

Notes:

* Genome coordinates were converted from version U00096.2 to version U00096.3 [here](https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12) 

```{r ht-tss-conway}
## Code executed just once, in order to export promoters coordinates and convert them through Ecocyc
##---------------------------------------------------------------------------------
# file <- ht_tss_dataset_list$DS0006$file
# read_excel(paste0(ht_collection_dir, "/coleccion_TSS/author_files/", file), sheet = "table s2 Promoters", skip = 2) %>%
#   dplyr::select(...3, STRAND) %>%
#   dplyr::mutate(pos_1 = gsub("^[PSIAS]{1,2}-", "", ...3), strand = STRAND) %>%
#   dplyr::select(strand, pos_1) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Conway_TSS_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)

## Conversion Conway_TSS_parsed.tsv -> Conway_TSS_parsed_converted.tsv

DS0006 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Conway_TSS_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::distinct() %>%
  dplyr::mutate(start = pos_1, stop = pos_1) 

```


### Ju

Reference article: [Ju et al., 2019](https://doi.org/10.1038/s41564-019-0500-z)

Conditions:

* exponential phase
* stationary phase


```{r ht-tss-ju}
file <- "41564_2019_500_MOESM3_ESM.xlsx"
DS0004_DS0005 <- read_excel(paste0(ht_collection_dir, "/coleccion_TSS/author_files/source_files/", file), skip = 1)  %>%
  dplyr::rename(pos_1 = TSS_site, 
                strand = TSS_direction) %>%
  dplyr::mutate(start = pos_1, stop = pos_1) %>%
  dplyr::select(-"TSS_sequence(5'_3')") %>%
  dplyr::distinct()

##---------------------------------------------------------------------------------
DS0004 <- DS0004_DS0005 %>%
  dplyr::filter(!is.na(Detected_in_log_phase)) %>%
  dplyr::select(-contains("Detected"))

DS0005 <- DS0004_DS0005 %>%
  dplyr::filter(!is.na(Detected_in_stationary_phase)) %>%
  dplyr::select(-contains("Detected"))
```


### Morett

Reference articles: 

1. [Mendoza-Vargas et al., 2009](https://doi.org/10.1371/journal.pone.0007526)
2. [Salgado et al., 2013](https://pubmed.ncbi.nlm.nih.gov/23203884/)


Conditions/methodology:

* 454 data (1)
* RACE data (1)
* 5 tri or monophosphate enrichment (1, 2)

Notes:

* Files were downloaded on 2021/03/19 [here](http://regulondb.ccg.unam.mx/highthroughputdatasetssearch?term=all)
* Genome coordinates were converted from version U00096.2 to version U00096.3 [here](https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12)
* 3 datasets files formatted differently:
  * One file has TSS left and right positions, as well as "max frequency" position.It also has "orientation" information.
  * Two files have gene's left and right coordinates, and relative position of the TSS to the gene (unique position is calculated depending on strand). These files don't have relative orientation information


```{r ht-tss-morett}
## Datasets 229 and 230: promoter position = gene start + relative orientation
## Dataset 225 (phosphate): promoter position = max_frequency column
##---------------------------------------------------------------------------------

## Code executed just once, in order to export promoters coordinates and convert them through Ecocyc
##---------------------------------------------------------------------------------
# file_454 <- "HTRI00000229_HighThroughputDataSets.txt"
# read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/", file_454), comment.char = "#", header = F) %>%
#   dplyr::rename(strand = V9, left = V4, right = V5, relative_pos = V11) %>%
#   dplyr::select(strand, left, right, relative_pos) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_454_TSS_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)
# 
# file_RACE <- "HTRI00000230_HighThroughputDataSets.txt"
# read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/", file_RACE), comment.char = "#", header = F) %>%
#   dplyr::rename(strand = V9, left = V4, right = V5, relative_pos = V11) %>%
#   dplyr::select(strand, left, right, relative_pos) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_RACE_TSS_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)

# file_phospate <- "HTRI00000225_HighThroughputDataSets.txt"
# read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/", file_phospate), comment.char = "#", header = F) %>%
#   dplyr::rename(start = V4, stop = V5, pos_1 = V6, strand = V9) %>%
#   dplyr::select(strand, start, stop, pos_1) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_phosphate_TSS_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)

## Conversion of genome coordinates through Biocyc: https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12
## Conversion Morett_454_TSS_parsed.tsv -> Morett_454_TSS_parsed_converted.tsv
## Conversion Morett_RACE_TSS_parsed.tsv -> Morett_RACE_TSS_parsed_converted.tsv
## Conversion Morett_phosphate_TSS_parsed.tsv -> Morett_phosphate_TSS_parsed_converted.tsv

DS0009 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_454_TSS_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::mutate(pos_1 = ifelse(strand == "forward", left + relative_pos, ifelse(strand == "reverse", right - relative_pos, NA)),
                start = pos_1, 
                stop = pos_1,
                strand = ifelse(strand == "reverse", "-", ifelse(strand == "forward", "+", NA))) %>%
  dplyr::distinct() 

DS0010 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_RACE_TSS_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::mutate(pos_1 = ifelse(strand == "forward", left + relative_pos, ifelse(strand == "reverse", right - relative_pos, NA)),
                start = pos_1, 
                stop = pos_1,
                strand = ifelse(strand == "reverse", "-", ifelse(strand == "forward", "+", NA))) %>%
  dplyr::distinct()

DS0011 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Morett_phosphate_TSS_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::mutate(strand = ifelse(strand == "reverse", "-", ifelse(strand == "forward", "+", NA))) %>%
  dplyr::distinct()
  

```

### Palsson

Reference article: [Cho et al., 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3923258/)

Conditions:

* Exponential phase (from [2012 paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3415461/)*)
* Glutamine as source of nitrogen
* Heat shock
* Stationary phase


Notes:

* **Pending curation**

Notes

* Genome coordinates were converted from version U00096.2 to version U00096.3 [here](https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12)


```{r pallson-ht-tss}
##---------------------------------------------------------------------------------
file <- "Palsson_2014_converted.tsv"
Palsson_all <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/", file), sep = "\t", header = TRUE, stringsAsFactors = F) %>%
  tidyr::separate_rows(Conditions, sep = "") %>%
  dplyr::mutate_all(dplyr::na_if,"") %>%
  dplyr::filter(!is.na(Conditions)) %>%
  dplyr::rename(strand = Strand, pos_1 = TSS.position, total_reads = TSS.reads, condition = Conditions) %>%
  dplyr::mutate(start = pos_1, stop = pos_1)

##---------------------------------------------------------------------------------
DS0013 <- Palsson_all %>%
  dplyr::filter(condition == "E")

DS0014 <- Palsson_all %>%
  dplyr::filter(condition == "G")

DS0015 <- Palsson_all %>%
  dplyr::filter(condition == "H")

DS0016 <- Palsson_all %>%
  dplyr::filter(condition == "S")

```

### Storz 

Reference article: [Thomason et al., 2014](https://doi.org/10.1128/jb.02096-14)

Conditions:

* LB growth medium, OD = 2.0 (stationary phase)
* Minimum growth medium, OD = 0.4 (exponential phase)
* LB growth medium, OD = 0.4 (exponential phase)

Notes:

* Supplementary file downloaded on 2021/11/02 [here](https://journals.asm.org/doi/suppl/10.1128/JB.02096-14/suppl_file/zjb999093409sd1.xlsx)
* Genome coordinates were converted from version U00096.2 to version U00096.3 [here](https://biocyc.org/ECOLI/map-seq-coords-form?chromosome=COLI-K12) 

```{r ht-tss-storz}
## Code executed just once, in order to export promoters coordinates and convert them through Ecocyc
##---------------------------------------------------------------------------------
# file <- "zjb999093409sd1.xlsx
# read_excel(paste0(ht_collection_dir, "/coleccion_TSS/author_files/", file), sheet = "TSS Map MasterTable", skip = 2) %>%
#   dplyr::rename(pos_1 = Pos, strand = Strand, condition = Condition) %>%
#   dplyr::select(pos_1, strand, condition, detected) %>%
#   write.table(file = paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Storz_TSS_parsed.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)

## Conversion Storz_TSS_parsed.tsv -> Storz_TSS_parsed_converted.tsv

DS0001_DS0002_DS0003 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/intermediate_files/Storz_TSS_parsed_converted.tsv"), comment.char = "#", header = T) %>%
  dplyr::distinct() %>%
  dplyr::mutate(start = pos_1, stop = pos_1) 

##---------------------------------------------------------------------------------
DS0001 <- DS0001_DS0002_DS0003 %>%
  dplyr::filter(condition == "LB_2.0" & detected == 1) %>%
  dplyr::select(-condition, -detected)

DS0002 <- DS0001_DS0002_DS0003 %>%
  dplyr::filter(condition == "M63_0.4" & detected == 1) %>%
  dplyr::select(-condition, -detected)

DS0003 <- DS0001_DS0002_DS0003 %>%
  dplyr::filter(condition == "LB_0.4" & detected == 1) %>%
  dplyr::select(-condition, -detected)
```

### Wade

Data not published, refer to [this article](https://doi.org/10.1007/978-1-62703-730-3_1) for details on methods

Conditions:

* M9 + 0.2% glycerol, cells grown with shaking at 30Â°C

Notes:

* TSSs queried from private database on 2021-09-21

```{r wade-tss}
##---------------------------------------------------------------------------------
file <- "Wade_promoters_21-09-21.tsv"
DS0012 <- read.delim(paste0(ht_collection_dir, "/coleccion_TSS/author_files/source_files/", file), comment.char = "#", header = T) %>%
  dplyr::mutate(pos_1 = start) %>%
  dplyr::select(start, stop, strand, pos_1) %>%
  dplyr::distinct() 

```

### Yan 

Reference paper: [Yan et al., 2018](10.1038/s41467-018-05997-6)

Conditions:

* M9 (minimal medium)
* Rich

Notes:

* HT-inferred TSSs from PacBio long read data.

```{r ht-tss-yan}
##---------------------------------------------------------------------------------
DS0007 <- read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/M9_RegulonDB_TU_definedEnd.txt"), stringsAsFactors = F) %>% 
  rbind.data.frame(read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/M9_RegulonDB_TU_longestRead.txt"), stringsAsFactors = F)) %>%
  dplyr::rename(start = V1, stop = V2, strand = V3) %>%
  dplyr::mutate(pos_1 = ifelse(strand == "-", stop, start), stop = pos_1, start = pos_1) %>%
  dplyr::select(start, stop, strand, pos_1) %>%
  dplyr::distinct()

##---------------------------------------------------------------------------------
DS0008 <- read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/Rich_RegulonDB_TU_definedEnd.txt"), stringsAsFactors = F) %>% 
  rbind.data.frame(read.table(paste0(ht_collection_dir, "/coleccion_TUs/author_files/source_files/Rich_RegulonDB_TU_longestRead.txt"), stringsAsFactors = F)) %>%
  dplyr::rename(start = V1, stop = V2, strand = V3) %>%
  dplyr::mutate(pos_1 = ifelse(strand == "-", stop, start), stop = pos_1, start = pos_1) %>%
  dplyr::select(start, stop, strand, pos_1) %>%
  dplyr::distinct()
```

## Uniformization

* Generate unique IDs for TSSs
* Write one file per dataset

```{r write-files}
tss_dir <- paste0(ht_collection_dir, "/coleccion_TSS/uniformized_data")
datasets_list <- list()

## Uniformize format, add TSS ID column, write file
for(ds_id in names(ht_tss_dataset_list)) {
  df <- get(ds_id) %>%
    dplyr::arrange(start) %>%
    dplyr::mutate(id = paste0("TSS_", ds_id, "_", dplyr::row_number())) %>%
    dplyr::mutate(chromosome = "NC_000913.3")

  write.table(df %>% dplyr::select(chromosome, start, stop, id, pos_1, strand),
              file = paste0(tss_dir, "/", ds_id, ".tsv"),
              sep = "\t", quote = F, col.names = T, row.names = F)
  
  datasets_list[[ds_id]] <- df
}

```

* Write a summary file for TSS datasets

```{r write-tss-table}
ht_tss_stats <- ht_tss_metadata %>%
  dplyr::rowwise() %>%
  dplyr::mutate(tss_number = nrow(get(dataset_id))) %>%
  dplyr::select(dataset_id, pmid, author, condition, tss_number)

write.table(ht_tss_stats, file = paste0(tss_dir, "/TSS_datasets.tsv"), sep = "\t", quote = F, col.names = T, row.names = F)
```

## Results

### Summary of datasets

```{r stats-tss}
DT2 <- DT::datatable(ht_tss_stats, options = list(dom = '', pageLength = 20)) %>% DT::formatRound('tss_number', digits=0)
DT2

total_tss <- nrow(bind_rows(datasets_list))

```

Total objects: `r total_tss`

### Distribution

```{r fig-stats-tss-1}
ht_tss_stats_legends <- ht_tss_stats  %>%
  dplyr::mutate(first_author = str_split(author, ",")[[1]][1])

ggplot(ht_tss_stats_legends, aes(x = dataset_id, y = tss_number, fill = first_author)) +
  geom_col(size = 2) +
  # scale_fill_manual(values = RColorBrewer::brewer.pal(7, "Pastel2")) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), text = element_text(size = 14)) +
  ylab("Number of TSSs") +
  xlab("Dataset ID") +
  labs(title = "Number of TSSs per dataset")
  
```

### Distance to closest gene

```{r fig-stats-tss-3, fig.height = 24}
for(ds_id in names(ht_tss_dataset_list)) {
  df <- get(ds_id) %>%
    dplyr::mutate(close = EcoliGenes::closest_gene(pos_1, strand)) %>%
    dplyr::mutate(distance = as.numeric(paste0(strand, 1)) * (pos_1 - close))

  assign(ds_id, df)
}

tss_distance_plots <- list()
for(ds_id in names(ht_tss_dataset_list)) {
  df <- get(ds_id) 
  author  <- ht_tss_dataset_list[[ds_id]]$author
  condition  <- ht_tss_dataset_list[[ds_id]]$condition
  
  gg <- ggplot(df, aes(distance)) + 
	  geom_histogram(binwidth = 10) + 
    geom_vline(aes(xintercept = 0, color = "red")) +
    geom_vline(aes(xintercept = -100, color = "red")) +
    theme(legend.position = "none", text = element_text(size = 14)) +
	  scale_x_continuous(limits = c(-1000,500)) +
    ylab("Number of TSSs") +
    xlab("Distance to closest gene start") +
    labs(title = ds_id, subtitle = paste0("Author: ", author, ", condition: ", condition))
  tss_distance_plots[[ds_id]] <- gg
}
do.call("grid.arrange", c(tss_distance_plots, nrow = length(ht_tss_dataset_list)))
```

